include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+
+    /// Logs the "status.measurement_done" event on std::clog.
+    void on_status_measurement_done(const events::StatusMeasurementDoneEvent &event) override;
+
+    /// Logs the "status.report_close" event on std::clog.
+    void on_status_report_close(const events::StatusReportCloseEvent &event) override;
+
+    /// Logs the "status.report_create" event on std::clog.
+    void on_status_report_create(const events::StatusReportCreateEvent &event) override;
+
+    /// Logs the "status.resolver_lookup" event on std::clog.
+    void on_status_resolver_lookup(const events::StatusResolverLookupEvent &event) override;
+
+    /// Logs the "status.started" event on std::clog.
+    void on_status_started(const events::StatusStartedEvent &event) override;
+
+    /// Logs the "status.update.performance" event on std::clog.
+    void on_status_update_performance(const events::StatusUpdatePerformanceEvent &event) override;
+
+    /// Logs the "status.update.websites" event on std::clog.
+    void on_status_update_websites(const events::StatusUpdateWebsitesEvent &event) override;
+
+    /// Logs the "task_terminated" event on std::clog.
+    void on_task_terminated(const events::TaskTerminatedEvent &event) override;
+
+    ~NoisyRouter() noexcept override;
+};
+
+} // namespace routers
+
+/// Manages the lifecycle of a nettest. This API mirrors as closely as
+/// possible to FFI API provided by <measurement_kit/ffi.h>.
+class Nettest {
+  public:
+    // C++ object model
+    // ````````````````
+
+    /// Creates a nettest instance. This will not start the actual test. You
+    /// need to use one of the start_XXX() methods to do that.
+    Nettest() noexcept;
+
+    /// Explicitly deleted copy constructor.
+    Nettest(const Nettest &) noexcept = delete;
+
+    /// Explicitly deleted copy assignment.
+    Nettest &operator=(const Nettest &) noexcept = delete;
+
+    /// Explicitly deleted move constructor.
+    Nettest(Nettest &&) noexcept = delete;
+
+    /// Explicitly deleted move assignment.
+    Nettest &operator=(Nettest &&) noexcept = delete;
+
+    /// Wait for nettest to terminate and destroy resources.
+    ~Nettest() noexcept;
+
+    // Starting a nettest
+    // ``````````````````
+
+    /// Starts a "CaptivePortal" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_captive_portal(settings::CaptivePortalSettings settings, routers::Router *router) noexcept;
```

Comment:
Why a different function for each nettest? Mainly because I need to initialise the internal JSON setting with specific variables (like the test name and other nettest specific settings) before running. An alternative design would have been to allow for serialisation of nettests to string and then to have a single start method receiving a string. Yet, I do not want to expose too much that we're using JSON internally. So, I decided to put all the logic in this class and to have different start methods. It could also be that we have a specific class for each nettest. I think this design is quite equivalent to the one here, but maybe I am missing some bits. Feedback appreciated.